### PR TITLE
Preload the preloaded subincludes in files involved in normal (non-preloaded) subincludes 

### PR DESCRIPTION
**Pull request by Tatskaari**
This means that we can be more specific about where we don't preload subincludes. Before any package involved in any kind of subinclude would be skipped. Now it's only if that package is part of the preloading.

### CODE REVIEW

File: src/build/BUILD [hunk 1]
```diff
@@ -36,3 +36,18 @@ go_test(
         "//src/fs",
     ],
 )
+
+go_test(
```
Comment (Tatskaari) [inline src/build/BUILD:40]: oops

File: test/preloaded_subinc/test_repo/src/build_defs/BUILD_FILE [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# This should be available to us via preloads even though this file is involved in a subinclude
+blah = sh_test
```
Comment (Tatskaari) [inline test/preloaded_subinc/test_repo/src/build_defs/BUILD_FILE:2]: This used to fail but we're more specific about when we don't preload now

File: src/parse/parse_step.go [hunk 3]
```diff
@@ -45,22 +45,22 @@ func parse(state *core.BuildState, label, dependent core.BuildLabel, forSubinclu
 
 	// Ensure that all the preloaded targets are built before we sync the package parse. If we don't do this, we might
 	// take the package lock for a package involved in a subinclude, and end up in a deadlock
-	if !forSubinclude {
+	if !mode.IsPreload() {
 		state.WaitForPreloadedSubincludeTargetsAndEnsureDownloaded()
 	}
```
Comment (Tatskaari) [inline src/parse/parse_step.go:50]: Being more specific here

File: src/core/state.go [hunk 4]
```diff
@@ -631,7 +646,7 @@ func (state *BuildState) WaitForPreloadedSubincludeTargetsAndEnsureDownloaded()
 			// Queue them up asynchronously to feed the queues as quickly as possible
 			wg.Add(1)
 			go func(inc BuildLabel) {
-				state.WaitForTargetAndEnsureDownload(inc, OriginalTarget)
+				state.WaitForTargetAndEnsureDownload(inc, OriginalTarget, true)
 				wg.Done()
 			}(inc)
```
Comment (Tatskaari) [inline src/core/state.go:651]: this sets the preload flags

File: src/core/state.go [hunk 5]
```diff
@@ -23,6 +23,21 @@ import (
 	"github.com/thought-machine/please/src/process"
 )
 
+type ParseMode uint8
+
+const (
+	ParseModeForSubinclude ParseMode = 1 << iota
```
Comment (peterebden) [inline src/core/state.go:29]: how about adding a `ParseModeNormal = 0` constant? We often call things with 0 which is a little unclear at the call site.

File: src/core/state.go [hunk 6]
```diff
@@ -23,6 +23,21 @@ import (
...
+	ParseModeForPreload
```
Comment (peterebden) [inline src/core/state.go:30]: do you need `ParseModeForPreload ParseMode`? I feel I've been caught out by this before

**Review by peterebden — APPROVED** [review]

File: src/core/state.go (hunk 6 above)
Comment (Tatskaari) [reply src/core/state.go:30]: It seems to compile fine so I assume not?

**Pull request merged**